src/Compilers/Core/Portable/Syntax/GreenNode.cs

```diff
@@ -644,47 +644,49 @@ public void WriteTo(System.IO.TextWriter writer)
         protected internal void WriteTo(TextWriter writer, bool leading, bool trailing)
         {
             // Use an actual Stack so we can write out deeply recursive structures without overflowing.
-            var stack = new Stack<(GreenNode node, bool leading, bool trailing)>();
+            var stack = ArrayBuilder<(GreenNode node, bool leading, bool trailing)>.GetInstance();
             stack.Push((this, leading, trailing));
 
             // Separated out stack processing logic so that it does not unintentionally refer to 
             // "this", "leading" or "trailing.
-            ProcessStack(writer, stack);
-        }
+            processStack(writer, stack);
+            stack.Free();
 
-        private static void ProcessStack(TextWriter writer,
-            Stack<(GreenNode node, bool leading, bool trailing)> stack)
-        {
-            while (stack.Count > 0)
+            static void processStack(
+                TextWriter writer,
+                ArrayBuilder<(GreenNode node, bool leading, bool trailing)> stack)
```

Comment:
1. thanks for making static.  that's very helpful.
2. could you consider a 'return;' statement before the local function?  That helps make it clear that there is no other substantive procedural code that needs to run after stack.free is called.